core/src/main/java/com/google/cloud/sql/core/SqlAdminApiService.java

```diff
@@ -0,0 +1,230 @@
+      } catch (IOException ex) {
+        throw addExceptionContext(ex, "An exception occurred while fetching IAM auth token:",
+            instanceName);
+      }
+    }
+    GenerateEphemeralCertResponse response;
+    try {
+      response = apiClient.connect()
+          .generateEphemeralCert(instanceName.getProjectId(), instanceName.getInstanceId(), request)
+          .execute();
+    } catch (IOException ex) {
+      throw addExceptionContext(ex,
+          String.format("[%s] Failed to create ephemeral certificate for the Cloud SQL instance.",
+              instanceName.getConnectionName()), instanceName);
+    }
+
+    // Parse the certificate from the response.
+    Certificate ephemeralCertificate;
+    try {
+      ephemeralCertificate = createCertificate(response.getEphemeralCert().getCert());
+    } catch (CertificateException ex) {
+      throw new RuntimeException(String.format(
+          "[%s] Unable to parse the ephemeral certificate for the Cloud SQL instance.",
+          instanceName.getConnectionName()), ex);
+    }
+
+    return ephemeralCertificate;
+  }
+
+  /**
+   * Checks for common errors that can occur when interacting with the Cloud SQL Admin API, and adds
+   * additional context to help the user troubleshoot them.
+   *
+   * @param ex exception thrown by the Admin API request
+   * @param fallbackDesc generic description used as a fallback if no additional information can be
+   * provided to the user
+   */
+  private RuntimeException addExceptionContext(IOException ex, String fallbackDesc,
+      CloudSqlInstanceName instanceName) {
+    // Verify we are able to extract a reason from an exception, or fallback to a generic desc
+    GoogleJsonResponseException gjrEx =
+        ex instanceof GoogleJsonResponseException ? (GoogleJsonResponseException) ex : null;
+    if (gjrEx == null || gjrEx.getDetails() == null || gjrEx.getDetails().getErrors() == null
+        || gjrEx.getDetails().getErrors().isEmpty()) {
+      return new RuntimeException(fallbackDesc, ex);
+    }
+    // Check for commonly occurring user errors and add additional context
+    String reason = gjrEx.getDetails().getErrors().get(0).getReason();
+    if ("accessNotConfigured".equals(reason)) {
+      // This error occurs when the project doesn't have the "Cloud SQL Admin API" enabled
+      String apiLink = "https://console.cloud.google.com/apis/api/sqladmin/overview?project="
+          + instanceName.getProjectId();
+      return new RuntimeException(String.format(
+          "[%s] The Google Cloud SQL Admin API is not enabled for the project \"%s\". Please "
+              + "use the Google Developers Console to enable it: %s",
+          instanceName.getConnectionName(), instanceName.getProjectId(), apiLink), ex);
+    } else if ("notAuthorized".equals(reason)) {
+      // This error occurs if the instance doesn't exist or the account isn't authorized
+      // TODO(kvg): Add credential account name to error string.
```

Comment:
Let's also strip out stale TODOs -- either by deleting them or making an issue and fixing them.